docs/routing.md

```diff
@@ -0,0 +1,266 @@
+    getNextPhase --> NextPhaseIs{determine next phase}
+
+    NextPhaseIs --> |none| NextIsFS[filesystem]
+    NextPhaseIs --> |filesystem| NextIsRW[rewrite]
+    NextPhaseIs --> |rewrite| NextIsRS[resource]
+    NextPhaseIs --> |resource| NextIsMS[miss]
+    NextPhaseIs --> |all others| NextIsMS
+```
+
+### Checking Source Routes
+
+For every source route in a phase, we need to check for a match and process it accordingly.
+
+#### Source Matching
+
+The first step in this part of the process is checking the value for the route matches up with the current path. At the same time, this also extracts any dynamic parameters.
+
+Checking whether a route is a match involves verifying that the request method is the same, the correct fields are or aren't specified, and the current path name matches.
+
+If the route was not a match, we simply skip it (i.e. it doesn't involve the current request).
+
+#### Overrides
+
+In different phases, different things may be applied to our response object, or modify the current requests details. The first thing we check for is whether a (matching) source route wants to override the current headers and status code and, if so, reset them.
+
+#### Locale
+
+To support internationalization redirects, the Vercel build output configuration contains source routes that map locales to redirects for both [sub-path](https://nextjs.org/docs/pages/building-your-application/routing/internationalization#sub-path-routing) and [domain](https://nextjs.org/docs/pages/building-your-application/routing/internationalization#domain-routing) routing. Dealing with these is a little challenging due to the format of some of the records and, as such, has involved custom behaviour.
+
+When we encounter a list of a locale redirects, we have to check for any cookie value that is provided as a way to override the request's `Accept-Language` header. Both the cookie and the header have to be parsed into a form that represents the locales, sorted by [quality](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language#directives). Vercel considers the locale in the cookie to take precendence over the header.
+
+If we find a locale that matches the current request, we apply a redirect for it and exit the routing process. At the same time, whenever we encounter a source route with a locale property, we store the list of locales for later use when dealing with edge cases.
+
+The custom behaviour that was referred to above comes into play when processing sub-path locale source routes. The `src` value that is used to match the request appears to only be for `/`, instead of `^/$`. This means that we have to check if the `src` value is a regular expression, and if not, handle it another way, to ensure that we are on the root. This is because automatic locale detection should [only](https://nextjs.org/docs/pages/building-your-application/routing/internationalization#automatic-locale-detection) occur at the root of the site.
```

Comment:
very nit 😅 

```suggestion
The custom behaviour that was referred to above comes into play when processing sub-path locale source routes. The `src` value that is used to match the request appears to only be for `/`, instead of `^/$`. This means that we have to check if the `src` value is a regular expression, and if not, handle it another way, to ensure that we are on the root. This is because automatic locale detection should [only occur at the application's root](https://nextjs.org/docs/pages/building-your-application/routing/internationalization#automatic-locale-detection).
```